File: bin/build-css-bundles.py
#!/usr/bin/env python3
from pathlib import Path
import re


ROOT = Path(__file__).resolve().parents[1]
SOURCE_DIR = ROOT / "assets" / "scss"
OUTPUT_DIR = ROOT / "build" / "assets" / "css"

BUNDLES = {
	"vkbm-frontend.min.css": [
		"variables.scss",
		"common.scss",
	],
	"vkbm-auth.min.css": [
		"variables.scss",
		"buttons.scss",
		"alert.scss",
		"auth-forms.scss",
	],
	"vkbm-editor.min.css": [
		"variables.scss",
		"utility.scss",
		"buttons.scss",
		"alert.scss",
		"auth-forms.scss",
		"admin-editor-fixes.scss",
		"common.scss",
	],
	"vkbm-admin.min.css": [
		"variables.scss",
		"utility.scss",
		"buttons.scss",
		"admin-notice.scss",
		"admin-table.scss",
		"admin-schedule.scss",
		"admin-provider-settings.scss",
		"admin-shift-editor.scss",
		"admin-shift-bulk-create.scss",
		"admin-shift-dashboard.scss",
		"admin-service-menu-quick-edit.scss",
		"admin-post-order.scss",
		"admin-term-order.scss",
		"admin-style-guide.scss",
		"admin-core.scss",
		"common.scss",
	],
}


def strip_comments(css: str) -> str:
    return re.sub(r"/\*.*?\*/", "", css, flags=re.S)


def normalize(css: str) -> str:
    css = strip_comments(css)
    css = re.sub(r"@import\s+url\([^)]+\);\s*", "", css)
    css = re.sub(r"\s+", " ", css)
    css = re.sub(r"\s*([{}:;,])\s*", r"\1", css)
    return css.strip()


def main() -> None:
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    for output_name, sources in BUNDLES.items():
        parts = []
        for source in sources:
            path = SOURCE_DIR / source
            if not path.exists():
                raise SystemExit(f"Missing source: {path}")
            parts.append(path.read_text(encoding="utf-8"))
        bundled = normalize("\n".join(parts))
        (OUTPUT_DIR / output_name).write_text(bundled + "\n", encoding="utf-8")


if __name__ == "__main__":
    main()

File: bin/build-css-bundles.js
#!/usr/bin/env node
const fs = require( 'fs' );
const path = require( 'path' );
const sass = require( 'sass' );
const CleanCSS = require( 'clean-css' );

const ROOT = path.resolve( __dirname, '..' );
const SOURCE_DIR = path.join( ROOT, 'assets', 'scss' );
const OUTPUT_DIR = path.join( ROOT, 'build', 'assets', 'css' );

const BUNDLES = {
	// --vkbm--* カスタムプロパティの既定値（:root ブロック）はここだけに出力する。
	// 他のバンドルから variables.scss を外し、この1本にまとめることで :root の重複出力を
	// 防ぐ（issue #420）。Common_Styles::register_styles() が他バンドルの $deps にこの
	// ハンドルを指定し、enqueue_block_assets（フロント・管理画面・ブロックエディターの
	// キャンバス iframe いずれでも発火する）で確実に読み込ませる。
	'vkbm-variables.min.css': [ 'variables.scss' ],
	'vkbm-frontend.min.css': [ 'common.scss' ],
	'vkbm-auth.min.css': [
		'buttons.scss',
		'alert.scss',
		'auth-forms.scss',
	],
	'vkbm-editor.min.css': [
		'utility.scss',
		'buttons.scss',
		'alert.scss',
		'auth-forms.scss',
		'admin-editor-fixes.scss',
		'common.scss',
	],
	'vkbm-admin.min.css': [
		'variables-admin.scss',
		'utility.scss',
		'buttons.scss',
		'alert.scss',
		'auth-forms.scss',
		'admin-notice.scss',
		'admin-table.scss',
		'admin-schedule.scss',
		'admin-provider-settings.scss',
		'admin-shift-editor.scss',
		'admin-shift-bulk-create.scss',
		'admin-shift-dashboard.scss',
		'admin-service-menu-quick-edit.scss',
		'admin-post-order.scss',
		'admin-term-order.scss',
		'admin-style-guide.scss',
		'admin-core.scss',
		'common.scss',
	],
};

function readScssFile( fileName ) {
	const fullPath = path.join( SOURCE_DIR, fileName );
	if ( ! fs.existsSync( fullPath ) ) {
		throw new Error( `Missing source: ${ fullPath }` );
	}
	const result = sass.compile( fullPath, {
		style: 'expanded',
		loadPaths: [ SOURCE_DIR ],
	} );
	return result.css;
}

function minifyCss( css ) {
	const result = new CleanCSS( { level: 2 } ).minify( css );
	if ( result.errors && result.errors.length ) {
		throw new Error( result.errors.join( '\n' ) );
	}
	return result.styles.trim();
}

function buildBundles() {
	fs.mkdirSync( OUTPUT_DIR, { recursive: true } );
	Object.entries( BUNDLES ).forEach( ( [ outputName, sources ] ) => {
		const parts = sources.map( readScssFile );
		const bundled = minifyCss( parts.join( '\n' ) );
		const outputPath = path.join( OUTPUT_DIR, outputName );
		fs.writeFileSync( outputPath, `${ bundled }\n`, 'utf8' );
	} );
}

buildBundles();
